### PR TITLE
📈 872 Implement Advanced Editor

### PR DESCRIPTION
**Pull request by xoscar**
This PR adds the first version of the advanced editor for the Tracetest selector query language.

## Changes

- Adds the new advanced editor component.

## Fixes

- https://github.com/kubeshop/tracetest/issues/872

## Checklist

- [x] tested locally
- [x] added new dependencies
- [ ] updated the docs
- [x] added a test

https://www.loom.com/share/6b13322e6b7d415f9beaaf0bdec34eaf

### CODE REVIEW

File: web/cypress/integration/TestDetails/ShowTestDetails.spec.ts [hunk 1]
```diff
@@ -25,6 +25,7 @@ describe('Show test details', () => {
     cy.visit(`http://localhost:3000/test/${testId}`);
 
     cy.get('[data-cy^=result-actions-button]').last().click();
+    cy.wait(5000);
```
Comment (xoscar) [inline web/cypress/integration/TestDetails/ShowTestDetails.spec.ts:28]: There is a race condition here. I'll be taking a deep look in the next PR

File: web/src/components/AdvancedEditor/AdvancedEditor.tsx [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+import useLint from './hooks/useLint';
+
+interface IProps {
+  testId: string;
+  runId: string;
+  value?: string;
+  onChange?(value: string): void;
+}
+
+const AdvancedEditor = ({testId, runId, onChange = noop, value = ''}: IProps) => {
```
Comment (xoscar) [inline web/src/components/AdvancedEditor/AdvancedEditor.tsx:17]: Main Editor component

File: web/src/components/AdvancedEditor/hooks/useAutoComplete.ts [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+import {useAppStore} from '../../../redux/hooks';
+import AssertionSelectors from '../../../selectors/Assertion.selectors';
+import {escapeString} from '../../../utils/Common';
+
+interface IProps {
+  testId: string;
+  runId: string;
+}
+
+const useAutoComplete = ({testId, runId}: IProps) => {
```
Comment (xoscar) [inline web/src/components/AdvancedEditor/hooks/useAutoComplete.ts:22]: Returns the handling function for the autocomplete extension

File: web/src/components/AssertionForm/AssertionForm.tsx [hunk 4]
```diff
@@ -114,6 +117,13 @@ const AssertionForm: React.FC<TAssertionFormProps> = ({
             You can decided if you want to use the wizard to create the span selector or the query language.
             `}
           />
+          {currentIsAdvancedSelector && (
```
Comment (xoscar) [inline web/src/components/AssertionForm/AssertionForm.tsx:120]: Displays a link to the reference documentation if advanced is selected

File: web/src/components/AssertionForm/AssertionFormSelector.tsx [hunk 5]
```diff
@@ -112,23 +113,18 @@ const AssertionFormSelector = ({
...
+    <S.AdvancedSelectorInputContainer>
+      <Form.Item
+        name="selector"
+        rules={[{required: true, message: 'The selector cannot be empty'}]}
+        validateTrigger={[]}
+        hasFeedback
+        help={isInvalidSelector ? 'Invalid selector' : ''}
+        validateStatus={isInvalidSelector ? 'error' : ''}
+      >
+        <AdvancedEditor runId={runId} testId={testId} />
```
Comment (xoscar) [inline web/src/components/AssertionForm/AssertionFormSelector.tsx:125]: Using the new editor

File: web/src/constants/AdvancedEditor.constants.ts [hunk 6]
```diff
@@ -0,0 +1,55 @@
+import {CompareOperatorSymbolMap, PseudoSelector} from './Operator.constants';
```
Comment (xoscar) [inline web/src/constants/AdvancedEditor.constants.ts:1]: Tokens and lists

File: web/src/selectors/Assertion.selectors.ts [hunk 7]
```diff
@@ -32,6 +32,13 @@ const AssertionSelectors = () => {
   return {
     selectAffectedSpanList,
     selectAttributeList: createSelector(selectAffectedSpanList, SpanSelectors.selectAffectedSpans, (spanList, affectedSpans) => spanList.flatMap(span => span.attributeList).concat(SpanAttributeService.getPseudoAttributeList(affectedSpans.length))),
+    selectAllAttributeList: createSelector(stateSelector, paramsSelector, (state, {testId, runId}) => {
```
Comment (xoscar) [inline web/src/selectors/Assertion.selectors.ts:35]: Returns the full list of attributes from the span list

File: web/src/utils/grammar/grammar.js [hunk 8]
```diff
@@ -0,0 +1,20 @@
+// This file was generated by lezer-generator. You probably shouldn't edit it.
+import {LRParser} from '@lezer/lr';
+
+export const parser = LRParser.deserialize({
```
Comment (xoscar) [inline web/src/utils/grammar/grammar.js:4]: Output from the `lezer/@generator`

File: web/src/utils/grammar/syntax.grammar [hunk 9]
```diff
@@ -0,0 +1,47 @@
+@precedence { list @left }
```
Comment (xoscar) [inline web/src/utils/grammar/syntax.grammar:1]: The tracetest lang description

File: web/src/components/AdvancedEditor/hooks/useAutoComplete.ts [hunk 10]
```diff
@@ -0,0 +1,104 @@
...
+  const {getState} = useAppStore();
+
+  const getAttributeList = useCallback(() => {
+    const state = getState();
+    const defaultList = AssertionSelectors.selectAllAttributeList(state, testId, runId);
```
Comment (jorgeepc) [inline web/src/components/AdvancedEditor/hooks/useAutoComplete.ts:27]: I don't think you need the state here. Does this work?
```ts
const defaultList = useAppSelector(state => AssertionSelectors.selectAllAttributeList(state, testId, runId));
```
Maybe the `getAttributeList` is not needed, I mean you can call the `useAppSelector` directly in your return fn, but not a big deal I'm ok with that.

File: web/src/components/AdvancedEditor/hooks/useLint.ts [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+  testId: string;
+  runId: string;
+}
+
+const useLint = ({runId, testId}: IProps): LintSource => {
+  const {getState} = useAppStore();
+
+  const getValidAttributeList = useCallback(() => {
+    const state = getState();
+    const attributeList = uniqBy(AssertionSelectors.selectAllAttributeList(state, testId, runId), 'key');
```
Comment (jorgeepc) [inline web/src/components/AdvancedEditor/hooks/useLint.ts:18]: Same here

**Review by jorgeepc — COMMENTED** [review]
Great work @xoscar 🔥 Added a couple of comments.

File: web/src/constants/AdvancedEditor.constants.ts [hunk 12]
```diff
@@ -0,0 +1,55 @@
...
+export const operatorList = [
+  {
+    label: CompareOperatorSymbolMap.EQUALS,
+    type: 'operatorKeyword',
+  },
+  {
+    label: CompareOperatorSymbolMap.CONTAINS,
+    type: 'operatorKeyword',
+  },
+];
```
[the remark this answers is not in the document]
Comment (jorgeepc) [reply web/src/constants/AdvancedEditor.constants.ts:37]: Nevermind, I confused this with the checks

File: web/src/utils/grammar/grammar.js [hunk 13]
```diff
@@ -0,0 +1,20 @@
+// This file was generated by lezer-generator. You probably shouldn't edit it.
```
[the remark this answers is not in the document]
Comment (xoscar) [reply web/src/utils/grammar/grammar.js:1]: Yeah, I just updated it to match our internal tsconfig/eslint rules but yes this is an output from the lezer/generator lib

File: web/src/constants/AdvancedEditor.constants.ts (hunk 12 above)
[the remark this answers is not in the document]
Comment (xoscar) [reply web/src/constants/AdvancedEditor.constants.ts:37]: That's only supported for checks at the moment, right now we only support contains and equals for selectors

File: web/src/components/AssertionForm/AssertionForm.tsx [hunk 14]
```diff
@@ -114,6 +117,13 @@ const AssertionForm: React.FC<TAssertionFormProps> = ({
...
+            <S.ReferenceLink>
+              <a href={ADVANCE_SELECTORS_DOCUMENTATION_URL} target="_blank">
+                Query Language Reference
```
[the remark this answers is not in the document]
Comment (xoscar) [reply web/src/components/AssertionForm/AssertionForm.tsx:123]: Yes, let's bring that up to the team tomorrow during planning

File: web/src/components/AdvancedEditor/hooks/useAutoComplete.ts (hunk 10 above)
Comment (xoscar) [reply web/src/components/AdvancedEditor/hooks/useAutoComplete.ts:27]: Yes, you are right, I had an async process that was causing the callback to update a lot of times which was re-instantiating the extension and creating a weird UX bug. Let me try it again

Comment (xoscar) [reply web/src/components/AdvancedEditor/hooks/useAutoComplete.ts:27]: So, what is happening here is that after moving that to an upper selector the callback changes every time the defaultList changes and causes an extensions change which is the reason for this blinking. 

2022-07-06 1 36 33 p m

Comment (xoscar) [reply web/src/components/AdvancedEditor/hooks/useAutoComplete.ts:27]: But the getState function never changes that's why the blinking doesn't happen when using that approach

**Review by jorgeepc — COMMENTED** [review]
Do you think it's possible to avoid the API call if the selector is not valid?
Screen Shot 2022-07-06 at 15 13 16

**Review by jorgeepc — APPROVED** [review]
🔥

**Pull request merged**